src/lib/geo/geo.cpp

```diff
@@ -262,6 +264,67 @@ int get_distance_to_line(struct crosstrack_error_s *crosstrack_error, double lat
 	return return_value;
 }
 
+int get_distance_to_segment(struct crosstrack_error_s *crosstrack_error, double lat_now, double lon_now,
```

Comment:
@xinnad , since this method is nearly identical to `get_distance_to_line()`, is there any reason not to just add the functional difference stated to the `get_distance_to_line()` function?
```
	// Unlike get_distance_to_line, this returns the distance to the nearest point on the segment. If the point
	// lies before the start or after the end we return the distance to the nearest endpoint. This is useful in
	// calcluating distances to the nearest point on a polygon when the point measured from lies outside.
```

Could the distance to the closest endpoint simply be broken out into another function?  I'm reluctant about adding duplicated code for the sake of just returning a distance to a point.  It feels like a bit of refactoring here could save code and improve maintainability. Does that make sense to you as well?

![image](https://user-images.githubusercontent.com/2497951/185966525-d6c190d4-dd0f-48e1-b063-af4c9508f48c.png)

Reply:
While a lot of the logic is similar I think its behavior warrants separate treatment since the question we're really asking is "if the projection of the point in question falls between the two points in the segment, return the projection. If the point's projection lies after the end, return the distance to the endpoint. If the point's projection lies before the start, return the distance to the start point". I considered modifying the original function but as it stands modification breaks existing code, since get_distance_to_line's current behavior is to return zero with the "after end" flag set, but projects onto an extension of the segment if the test coordinate lies "before start". We could move the "before start" check inside the original distance to line function, setting a flag but returning the same value as before. The new function would then check these flags to replace the distance returned by that to the distance to the nearest endpoint, but when I started down that track I was concerned about breaking existing code again. Happy to rework this a little since there is definitely a little duplicate logic, but the two functions behaved differently enough this seemed to make the code the most readable. Let me know what you think.